x-pack/auditbeat/module/system/login/login.go

```diff
@@ -0,0 +1,193 @@
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+const (
+	moduleName    = "system"
+	metricsetName = "login"
+
+	bucketName = "auditbeat.login.v1"
+
+	eventTypeEvent = "event"
+
+	// LoginRecordTypeBoot represents a system boot.
+	LoginRecordTypeBoot = "boot"
+	// LoginRecordTypeShutdown represents a system shutdown (halt or reboot).
+	LoginRecordTypeShutdown = "shutdown"
+	// LoginRecordTypeUserLogin represents a user login.
+	LoginRecordTypeUserLogin = "user_login"
+	// LoginRecordTypeUserLogout represents a user logout.
+	LoginRecordTypeUserLogout = "user_logout"
+)
+
+// LoginRecord represents a login record.
+type LoginRecord struct {
+	Utmp      Utmp
+	Type      string
+	PID       int
+	TTY       string
+	UID       int
+	Username  string
+	Hostname  string
+	IP        *net.IP
+	Timestamp time.Time
+	Origin    string
+}
+
+func (login LoginRecord) toMapStr() common.MapStr {
+	mapstr := common.MapStr{}
+
+	// Very useful for development
+	//mapstr.Put("utmp", fmt.Sprintf("%v", login.Utmp))
+
+	if login.TTY != "" {
+		mapstr.Put("tty", login.TTY)
+	}
+
+	return mapstr
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet collects login records from /var/log/wtmp.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config     Config
+	log        *logp.Logger
+	utmpReader *UtmpFileReader
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := defaultConfig
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
+	}
+
+	ms.utmpReader, err = NewUtmpFileReader(ms.log, bucket, config.UtmpFilePattern)
+	if err != nil {
+		return nil, err
+	}
+
+	return ms, nil
+}
+
+// Close cleans up the MetricSet when it finishes.
+func (ms *MetricSet) Close() error {
+	return ms.utmpReader.Close()
+}
+
+// Fetch collects any new login records from /var/log/wtmp. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	loginRecords, err := ms.utmpReader.ReadNew()
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+		return
+	}
+	ms.log.Debugf("%d new login records.", len(loginRecords))
+
+	for _, loginRecord := range loginRecords {
+		report.Event(ms.loginEvent(loginRecord))
+	}
+
+	// Save new state to disk
+	if len(loginRecords) > 0 {
+		err := ms.utmpReader.saveStateToDisk()
```

Comment:
Mentioning it mostly for awareness: if I'm not wrong, the Auditbeat `report.Event()`doesn't guarantee sending at least once. If Elasticsearch is down, for example, it will retry 3 times, and then drop the event. This means that we can potentially lose login events, despite saving the state. I think we could solve this after the Filebeat refactoring, but I suggest documenting this as a limitation for now.

Reply:
Hm, that's a shame. It guess this affects all metricsets, and we might lose process starts, socket connections, password changes, audit events, FIM events, etc.

The [internal queue](https://www.elastic.co/guide/en/beats/auditbeat/master/configuring-internal-queue.html) does not keep events until they can be sent (or it overflows)? That would seem desirable.

Reply:
It does, and one can also use spooling to disk to make the queue larger. But Filebeat offers an extra guarantee by waiting for the ACK and only afterwards update the state on disk. This is not currently possible in Auditbeat. That’s the difference I wanted to highlight.